cmd/ci-operator-configresolver/main.go

```diff
@@ -119,18 +148,18 @@ func main() {
 	if err != nil {
 		logrus.WithError(err).Fatal("failed go gather options")
 	}
-	if err := validateOptions(o); err != nil {
+	if err := validateOptions(&o); err != nil {
 		logrus.Fatalf("invalid options: %v", err)
 	}
 	level, _ := logrus.ParseLevel(o.logLevel)
 	logrus.SetLevel(level)
 
-	configAgent, err := agents.NewConfigAgent(o.configPath, agents.WithConfigMetrics(configresolverMetrics.ErrorRate))
+	configAgent, err := agents.NewConfigAgent(o.configPath, agents.WithConfigMetrics(configresolverMetrics.ErrorRate), agents.WithReleaseRepoPath(o.releaseRepoPath))
 	if err != nil {
 		logrus.Fatalf("Failed to get config agent: %v", err)
 	}
 
-	registryAgent, err := agents.NewRegistryAgent(o.registryPath, agents.WithRegistryMetrics(configresolverMetrics.ErrorRate), agents.WithRegistryFlat(o.flatRegistry))
+	registryAgent, err := agents.NewRegistryAgent(o.registryPath, agents.WithRegistryMetrics(configresolverMetrics.ErrorRate), agents.WithRegistryFlat(o.flatRegistry), agents.WithRegistryReleaseRepoPath(o.releaseRepoPath))
```

Comment:
I know this is not new, but with separate agents/watchers, can we get into a situation where a request is serviced after `configAgent` has reloaded the files but `registryAgent` hasn't yet (or the reverse)?

Reply:
The way the code is designed, we will run 2 identical watchers, but I don't think we will get into a situation where one is updated and the other not since they are programmed to respond to the same event.

Reply:
What if this happens?

- `configAgent` is updated to include a test which uses a new registry component
- request arrives here
- `registryAgent` is updated to include the new registry component

Reply:
When the new symlink destination will be created, both of the agents will respond in the same event and reload their configs once.

Reply:
Right, but they operate without knowledge of each other.  How do we prevent the situation I described?  Suppose we are updating to a version that adds both a registry component and a `ci-operator` configuration file which uses it.

Reply:
But both agents will be updated simultaneously.

Reply:
Our agents are essentially doing (independently):

```go
for {
    waitForChanges()
    newConfig = loadConfig()
}
```

- [`configAgent`](https://github.com/openshift/ci-tools/blob/d9519e391da9003ee75b5700e359a8bb9ed7aefb/pkg/load/agents/configAgent.go#L221-L243)
- [`registryAgent`](https://github.com/openshift/ci-tools/blob/master/pkg/load/agents/registryAgent.go#L129-L155)

Because there is no synchronization between them, the situation in https://github.com/openshift/ci-tools/pull/2805#discussion_r873811638 would be:

- both agents are notified of the change
- `configAgent` finishes `newConfig = loadConfig()`
- a request arrives
- `registryAgent` finishes `newConfig = loadConfig()`

The request will fail, since the configuration returned by `configAgent` references a registry component which the `registryAgent` does not (yet) know.

Reply:
They are doing it independently but the same event will trigger them. With some divine power, they shouldn't race :). Let me check what it takes to refactor the whole thing to use the same watcher.

Reply:
To be clear: this problem has always existed in `ci-operator-configresolver`.  It is also somewhat improved by the use of a single directory instead of multiple `ConfigMap` mounts.  Before we had _no guarantee_ about when the registry update would happen in relation to the configuration update.

But if we're making changes here in order to solve the configuration update problem (which is the goal of DPTP-2531), we should at least have a plan for what to do about this.

Reply:
Before those agents were watching two different things. Now the agents are watching the exact same thing. It is quite unlike that they will race. Either way, I think it should be better to have only 1 watcher instead. I will do the changes.